### PR TITLE
chore: upgrade keycloak dependency

### PR DESCRIPTION
**Pull request by ableuler**
/deploy

~(still work in progress, but I want a CI deployment)~ ready for review now!

### CODE REVIEW

File: .github/workflows/pull-request-test.yml [hunk 1]
```diff
@@ -112,7 +112,7 @@ jobs:
       if: failure()
       uses: ./actions/download-test-artifacts
       env:
-        RENKU_VALUES: ${{ secrets.CI_RENKU_VALUES }}
+        RENKU_VALUES: ${{ secrets.CI_RENKU_VALUES_1900 }}
```
Comment (aledegano) [inline .github/workflows/pull-request-test.yml:115]: I guess these values will be reverted before merging, right?

Comment (ableuler) [reply .github/workflows/pull-request-test.yml:115]: Well... I'd find it the cleanest solution to create new values files and put them into new secrets every time we have to change the values. But currently this would break some assumptions in our component repos, so for the time being I'll just use this for testing, revert back once the PR is ready, update the secret to make the PR pass (thereby breaking all other CI pipelines) and merge this PR asap.

Comment (aledegano) [reply .github/workflows/pull-request-test.yml:115]: ah, right! The classical conundrum of the global configuration... yeah, I don't see many solutions, the only other one would be to include a timestamp/version in these variables (e.g. CI_RENKU_VALUES_20210202), but then we'll have to deal with the garbage...
I like your approach better.

File: helm-chart/README.rst [hunk 2]
```diff
@@ -56,3 +56,43 @@ To run tests on the deployment, use
...
+- If they have not been explicitly defined (and thus autocreated through a helm hook), make sure to get the value
+  of get the `keycloak-postgres-password` from the `renku-keycloak-postgres` secret and add it to your values file
+  as `global.keycloak.postgresPassword.value`. Do the same for `keycloak-password` from the `keycloak-password-secret`
+  and add it as `global.keycloak.postgres.value`.
+
+- Delete the two secrets which need to be recreated as well as the keycloak StatefulSet:
+.. code-block:: bash
+
+   kubectl delete secrets -n <namespace> keycloak-password-secret renku-keycloak-postgres
+   kubectl delete statefulsets.apps -n <namespace> <release-name>-keyclo
```
Comment (aledegano) [inline helm-chart/README.rst:90]:
```suggestion
   kubectl delete statefulsets.apps -n <namespace> <release-name>-keycloak
```

Comment (aledegano) [inline helm-chart/README.rst:90]:
```suggestion
   kubectl delete statefulsets.apps -n <namespace> <release-name>-key
```

File: helm-chart/README.rst [hunk 3]
```diff
@@ -56,3 +56,43 @@ To run tests on the deployment, use
...
+- Before starting, make sure to check out `the values changelog for this upgrade<https://github.com/SwissDataScienceCenter/renku/blob/master/helm-chart/values.yaml.changelog.md#upgrading-to-renku-080-includes-breaking-changes>`_
+  and update your values file accordingly.
+
+- The upgrade of keycloak will perform an irreversible database migration. It is therefore recommended
+  to back up your postgres volume before performing this upgrade.
+
+- If they have not been explicitly defined (and thus autocreated through a helm hook), make sure to get the value
+  of get the `keycloak-postgres-password` from the `renku-keycloak-postgres` secret and add it to your values file
+  as `global.keycloak.postgresPassword.value`. Do the same for `keycloak-password` from the `keycloak-password-secret`
+  and add it as `global.keycloak.postgres.value`.
```
Comment (aledegano) [inline helm-chart/README.rst:84]:
```suggestion
  and add it as `global.keycloak.password.value`.
```

File: helm-chart/README.rst [hunk 4]
```diff
@@ -56,3 +56,43 @@ To run tests on the deployment, use
...
+`the keycloak chart docs<https://github.com/codecentric/helm-charts/tree/master/charts/keycloak#upgrading>`_
+and the `keycloak docs<https://www.keycloak.org/docs/latest/upgrading/>`_.
+
+- Before starting, make sure to check out `the values changelog for this upgrade<https://github.com/SwissDataScienceCenter/renku/blob/master/helm-chart/values.yaml.changelog.md#upgrading-to-renku-080-includes-breaking-changes>`_
+  and update your values file accordingly.
+
+- The upgrade of keycloak will perform an irreversible database migration. It is therefore recommended
+  to back up your postgres volume before performing this upgrade.
+
+- If they have not been explicitly defined (and thus autocreated through a helm hook), make sure to get the value
```
Comment (aledegano) [inline helm-chart/README.rst:81]: I would rephrase this to really highlight that the two passwords have to be backed up before deleting them and just mention that is not necessary later.
e.g.:
"[WARNING] Backup the secrets: `renku-keycloak-postgres.keycloak-postgres-password` and `keycloak-password-secret.keycloak-password` with the command 
[...]
This operation is not needed if the secrets were previously explicitly defined [...]"

File: helm-chart/renku/values.yaml [hunk 5]
```diff
@@ -116,14 +118,55 @@ ingress:
...
+  # renku-keycloak-postgres
+  # - DB_ADDR: database host                            value: <RELEASE-NAME>-postgresql
+  # - DB_DATABASE: database name for keycloak           value: global.keycloak.postgresDatabase
+  # - DB_USER: database username for keycloak database  value: global.keycloak.postgresUser
+  # - DB_PASSWORD: database password for keycloak       value: global.keycloak.postgresPassword.value
+  # keycloak-password-secret
+  # - KEYCLOAK_USER: keyclaok admin username            value: global.keycloak.user
+  # - KEYCLOAK_PASSWORD: keycloak admin password        value: global.keycloak.password.value
+
+  extraEnvFrom: |
```
Comment (aledegano) [inline helm-chart/renku/values.yaml:150]: Something's not very clear to me: the global value: `global.keycloak.postgresPassword.value` and the secret `keycloak-password-secret` are a redundant: I don't understand if both should be present, or either depending if the operators is performing an upgrade or a clean installation.
Perhaps we could clarify in the instructions?

File: helm-chart/README.rst (hunk 2 above)
Comment (ableuler) [reply helm-chart/README.rst:90]: oops - I've seen so many 7-letter namespaces (andreas, ci-1234) that I've come to believe that `<release-name>-keyclo` was generic...

File: helm-chart/README.rst (hunk 3 above)
Comment (ableuler) [reply helm-chart/README.rst]: 👍

File: helm-chart/renku/values.yaml (hunk 5 above)
Comment (ableuler) [reply helm-chart/renku/values.yaml:150]: The value of `global.keycloak.postgresPassword.value` is written into `keycloak-password-secret`. If `global.keycloak.postgresPassword.value` is missing, the actual password is autogenerated through helm. In this second case, we have to copy the password from the secret into the values file if we want the password to remain unchanged over the deletion of the secret and the subsequent `helm upgrade`.

File: helm-chart/README.rst (hunk 4 above)
Comment (ableuler) [reply helm-chart/README.rst:81]: 👍

File: helm-chart/README.rst (hunk 2 above)
Comment (ableuler) [reply helm-chart/README.rst]: fixed by querying the actual name first.

File: helm-chart/README.rst (hunk 4 above)
Comment (ableuler) [reply helm-chart/README.rst]: Tried to make things clearer.

File: helm-chart/renku/values.yaml (hunk 5 above)
Comment (ableuler) [reply helm-chart/renku/values.yaml:150]: I agree that the secrets handling could be documented better in general, but I'm not sure the instructions on how to upgrade the renku helm chart version is the best place for this.

Comment (aledegano) [reply helm-chart/renku/values.yaml:150]: ok, thanks for the explanation.

**Review by aledegano — DISMISSED** [review]

File: helm-chart/renku/values.yaml [hunk 6]
```diff
@@ -116,14 +118,55 @@ ingress:
...
+      emptyDir: {}
+
+
   ## Create a demo user in keycloak? Note that the password for the demo
   ## user must be queried from kubernetes (see the rendered NOTES.txt
   ## template which is shown after a successul deployment).
   createDemoUser: false
 
-  ## Spawn a keycloak instance
-  enabled: true
```
Comment (pameladelgado) [inline helm-chart/renku/values.yaml:125]: dont we need this anymore? what about deployments with external keycloak?

**Review by pameladelgado — COMMENTED** [review]
Thanks for doing this!
I haven't tried this yet but it looks good. I just have a couple of comments/questions.
Have you already tried updating keycloak in an existing deployment? otherwise I could do it.

Also, the changelog notes are clear enough to me.

File: helm-chart/renku/templates/_keycloak-clients-users.tpl [hunk 7]
```diff
@@ -46,7 +46,7 @@ Define clients and users for Keycloak
 [
   {{- if .Values.keycloak.createDemoUser -}}
   {
-    "username": "demo",
+    "username": "demo@datascience.ch",
```
[the remark this answers is not in the document]
Comment (ableuler) [reply helm-chart/renku/templates/_keycloak-clients-users.tpl:49]: I agree, not the smartest detail for a demo user. However, in this PR I merely fixt the fact that the email should be used as username for our current keycloak settings. Older keycloak versions still tolerated using a non-email username, but now this fails.

File: helm-chart/renku/values.yaml (hunk 6 above)
Comment (ableuler) [reply helm-chart/renku/values.yaml:125]: It's still there, just moved a bit.

File: helm-chart/README.rst [hunk 8]
```diff
@@ -56,3 +56,49 @@ To run tests on the deployment, use
...
+
+- Before starting, make sure to check out `the values changelog for this upgrade<https://github.com/SwissDataScienceCenter/renku/blob/master/helm-chart/values.yaml.changelog.md#upgrading-to-renku-080-includes-breaking-changes>`_
+  and update your values file accordingly.
+
+- The upgrade of keycloak will perform an **irreversible database migration**. It is therefore recommended
+  to **back up your postgres volume** before performing this upgrade.
+
+- **Warning: Persist keycloak-related secrets!**
+
+  If `global.keycloak.postgresPassword.value` and `global.keycloak.postgresPassword.value`
```
Comment (olevski) [inline helm-chart/README.rst:83]: @ableuler I think you mean to say here the postgress password and the keycloak password. But you have just the same thing repeated.

File: helm-chart/README.rst [hunk 9]
```diff
@@ -56,3 +56,49 @@ To run tests on the deployment, use
...
+  have not been explicitly defined in the values file (and thus have been autocreated by helm),
+  add them to the values file now.
+
+  * Get the `keycloak-postgres-password` from the `renku-keycloak-postgres` secret and add it as `global.keycloak.postgresPassword.value`.
```
Comment (olevski) [inline helm-chart/README.rst:87]: @ableuler this is probably just me... But I overlooked here that the revised section in my values file should be:
```
global:
  keycloack:
    postgressPassword:
      value: actual_password_value
```

And not 
```
global:
  keycloack:
    postgressPassword: actual_password_value
```

Not sure if it is worth adding this as an example here. I could not make a code suggestion that included a multiline code block inside of it.

**Review by olevski — APPROVED** [review]
@ableuler you have my approval for this. I tested on my dev environment and following your instructions I did not lose (what little data) I had in my keycloak.

The proposals for edits in the documentation are very minor so you can do them here and merge or do them in a subsequent PR.

Oh and a few reminders:
- update the branch because in the meantime some change occurred on master
- revert back the `secrets.CI_RENKU_VALUES_1900` values when appropriate

**Review by aledegano — DISMISSED** [review]

File: helm-chart/renku/templates/_keycloak-clients-users.tpl (hunk 7 above)
[the remark this answers is not in the document]
Comment (pameladelgado) [reply helm-chart/renku/templates/_keycloak-clients-users.tpl:49]: right, we could maybe add this as an optional parameter for the values file. What do you think? It could also be in another PR...

[the remark this answers is not in the document]
Comment (ableuler) [reply helm-chart/renku/templates/_keycloak-clients-users.tpl:49]: Yeah, I think the cleanest would be that if the admin want's to automatically create users, all details of those users have to be provided explicilty, no assumptions from the chart side about those users.

File: helm-chart/README.rst (hunk 8 above)
Comment (ableuler) [reply helm-chart/README.rst:83]: 👍

**Comment by ableuler** [comment]
@olevski thanks for the careful review - added all your suggestions.

**Review by olevski — APPROVED** [review]
